File: platform/core/messaging/src/main/kotlin/org/whiteprint/platform/core/messaging/inbox/EventInbox.kt
package org.whiteprint.platform.core.messaging.inbox

import org.whiteprint.platform.core.messaging.model.event.EventScope
import java.time.Instant

interface EventInbox {
    val eventId: Long
    val traceId: String
    val causationId: String?
    val occurredAt: Instant
    val issuer: String
    val producer: String
    val schemaVersion: String
    val partitionKey: Long
    val eventScope: EventScope
    val eventType: String
    val payload: ByteArray
    val payloadJson: String
    val metadataJson: String
    val status: EventInboxStatus
    val receivedAt: Instant
    val processedAt: Instant?
    val attemptCount: Int
    val lastAttemptedAt: Instant?
    val errorMessage: String?
}

File: platform/adapter/event/inbox/src/main/resources/mongo/eventInbox.js
// Collection: event_inbox
// Document structure reference

db.createCollection("event_inbox");

db.event_inbox.createIndex({ event_type: 1, status: 1 });

// Document example:
// {
//     _id:              Long,         // event_id
//     trace_id:         String,
//     causation_id:     String | null,
//     occurred_at:      Date,
//     issuer:           String,
//     producer:         String,
//     schema_version:   String,
//     partition_key:    Long,
//     event_scope:      String,       // EventScope enum
//     event_type:       String,
//     payload:          BinData,
//     payload_json:     String,
//     metadata_json:    String,
//     status:           String,       // RECEIVED | PROCESSING | COMPLETED | FAILED | DEAD
//     received_at:      Date,
//     processed_at:     Date | null,
//     attempt_count:    Int,
//     last_attempted_at: Date | null,
//     error_message:    String | null
// }
